libnetwork/drivers/bridge/port_mapping_linux.go

```diff
@@ -1,221 +1,319 @@
 package bridge
 
 import (
-	"bytes"
 	"context"
 	"errors"
 	"fmt"
 	"net"
 
 	"github.com/containerd/log"
+	"github.com/docker/docker/libnetwork/iptables"
 	"github.com/docker/docker/libnetwork/netutils"
+	"github.com/docker/docker/libnetwork/portallocator"
+	"github.com/docker/docker/libnetwork/portmapper"
 	"github.com/docker/docker/libnetwork/types"
-	"github.com/ishidawataru/sctp"
 )
 
-func (n *bridgeNetwork) allocatePorts(ep *bridgeEndpoint, reqDefBindIP net.IP, ulPxyEnabled bool) ([]types.PortBinding, error) {
-	if ep.extConnConfig == nil || ep.extConnConfig.PortBindings == nil {
-		return nil, nil
-	}
+type portBinding struct {
+	*types.PortBinding
+	proxy portmapper.UserlandProxy
+}
 
-	defHostIP := net.IPv4zero // 0.0.0.0
-	if reqDefBindIP != nil {
-		defHostIP = reqDefBindIP
+// allocatePorts takes cfg, the configuration for port mappings, selects host
+// ports when ranges are given, starts docker-proxy or its dummy to reserve
+// host ports, and sets up iptables NAT/forwarding rules as necessary. If
+// anything goes wrong, it will undo any work it's done and return an error.
+// Otherwise, the returned slice of portBinding has an entry per address
+// family (if cfg describes a mapping for 'any' host address, it's expanded
+// into mappings for IPv4 and IPv6, because that's how the mapping is presented
+// in 'inspect'). HostPort and HostPortEnd in each returned portBinding are set
+// to the selected and reserved port.
+func (n *bridgeNetwork) allocatePorts(
+	epAddrV4, epAddrV6 *net.IPNet,
+	cfg []types.PortBinding,
+	defHostIP net.IP,
+) (_ []*portBinding, retErr error) {
+	if len(defHostIP) == 0 {
+		defHostIP = net.IPv4zero
 	}
 
-	var containerIPv6 net.IP
-	if ep.addrv6 != nil {
-		containerIPv6 = ep.addrv6.IP
+	var containerIPv4, containerIPv6 net.IP
+	if epAddrV4 != nil {
+		containerIPv4 = epAddrV4.IP
 	}
-
-	pb, err := n.allocatePortsInternal(ep.extConnConfig.PortBindings, ep.addr.IP, containerIPv6, defHostIP, ulPxyEnabled)
-	if err != nil {
-		return nil, err
+	if epAddrV6 != nil {
+		containerIPv6 = epAddrV6.IP
 	}
-	return pb, nil
-}
 
-func (n *bridgeNetwork) allocatePortsInternal(bindings []types.PortBinding, containerIPv4, containerIPv6, defHostIP net.IP, ulPxyEnabled bool) ([]types.PortBinding, error) {
-	bs := make([]types.PortBinding, 0, len(bindings))
-	for _, c := range bindings {
-		bIPv4 := c.GetCopy()
-		bIPv6 := c.GetCopy()
-		// Allocate IPv4 Port mappings
-		if ok := n.validatePortBindingIPv4(&bIPv4, containerIPv4, defHostIP); ok {
```

Comment:
I renamed `validatePortBindingIPv4 ` to `configurePortBindingIPv4`, because that's what it does (and did). I don't think there's any less validation ... but, did you spot something specific?